README
This is ARPMONITOR, a simple tool to trace arp requests from/to your machine.
If you would like to monitor your network bus, or entire network ( if you dont use a switch or something like that ), just do a ifconfig <interface> PROMISC, to set the promiscuous mode of your network interface, and arpmonitor will automagically get all the traffic.
I did arpmonitor to help me in a research, in 1998, about arp packets, delays between switches ports and so on. 
NOTE: some var name and code was stolen from neped. since I didn`t suceed to contact the author, they are here...


Gleicon S. Moraes - gleicon@gmail.com

08/1998 - Initial release
05/2001 - I found the tarball, cleaned some points and posted to fm
09/2001 - Added promisc mode setting ( to use instead ifconfig promisc ), fixed some syslog errors, signal handling and command line options (see USAGE)
06/2018 - Fixed compiling for modern linux, moved from SOCK_PACKET to SOCK_RAW, fixed promisc mode. Updated email after finding this tarball, uploaded to github.

I am not maintaning it *actively* anymore, but, if you have a patch, or idea, drop me a line, I`ll try to implement/incorporate and maintain it useful.

I am lazy !




$version: 0.05


CHANGELOG

Gleicon S. Moraes - gleicon@terra.com.br

0.4 - 09/2001 - Added promisc mode setting ( to use instead ifconfig promisc ), fixed some syslog errors, signal handling , log to syslog or stderr, command line options (see USAGE), network code cleanup.

0.3 - 05/2001 - I found the tarball, cleaned some points and posted to fm

08/1998 - Initial release




main.h
/* main includes and defs */


#include <stdio.h> 
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/socket.h>
#include <sys/time.h>
#include <malloc.h>
#include <ctype.h>
#include <netinet/in.h> 
#include <arpa/inet.h>  
#include <netinet/tcp.h>
#include <net/if.h>   
#include <sys/ioctl.h>
#include <fcntl.h>
#include <time.h>
#include <signal.h>
#include <syslog.h>  
#include <sys/time.h>

#define OPTSTR "hpsei:"

#include "functions.h"


#define ETH_P_ARP       0x0806
#define MAX_PACK_LEN    2000
#define ETHER_HEADER_LEN 14
#define ARPREQUEST      1
#define ARPREPLY        2
#define perr(s) fprintf(stderr,"%s", s)

struct arp_struct {
	u_char  dst_mac[6];
        u_char  src_mac[6];
        u_short pkt_type;  
        u_short hw_type;   
        u_short pro_type;  
        u_char  hw_len;    
        u_char  pro_len;   
        u_short arp_op;    
        u_char  sender_eth[6];
        u_char  sender_ip[4]; 
        u_char  target_eth[6];
        u_char  target_ip[4]; 
        };
        
                                                  
union {
    	u_char full_packet[MAX_PACK_LEN];
        struct arp_struct arp_pkt;
        } a;
          
#define full_packet a.full_packet
#define arp_pkt a.arp_pkt

#define SET	0
#define UNSET	1




functions.c
/* general functions */


#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/socket.h>
#include <sys/time.h>
#include <malloc.h>
#include <ctype.h>  
#include <netinet/in.h>
#include <arpa/inet.h> 
#include <netinet/tcp.h>
#include <net/if.h>
#include <sys/ioctl.h>
#include <fcntl.h>
#include <time.h>
#include <errno.h>

#include <syslog.h>
#include <sys/time.h>
 

#define SET     0
#define UNSET   1


	
extern int rec;
extern char if_name[256];
extern int outputlog;


char *inetaddr ( u_int32_t ip ) {
  	struct in_addr in;
    	in.s_addr = ip;   
      	return inet_ntoa(in);
      	}

char *hwaddr (u_char * s) {
   	static char buf[30];
     	sprintf (buf, "%02X:%02X:%02X:%02X:%02X:%02X", s[0], s[1], s[2], s[3], s[4],s[5]);
        return buf;
        }
       
void logme (char *heh) {
	
	if (outputlog) {
		fprintf (stderr,"ArpMonitor: ");
		fprintf (stderr,"%s\n", heh);
	} else {
        	syslog (LOG_INFO, "%s", heh);
	}
}
       
void perr_syslog (char *heh) {
	if (outputlog) {
		fprintf (stderr,"ArpMonitor:");
		fprintf (stderr,"%s\n", heh);
	} else {
        	syslog (LOG_INFO, "error:%s", heh);
    	}
}

void promisc_mode (int rec, int status, char *if_name) {
	
	struct ifreq if_data;
	
	strncpy (if_data.ifr_name, if_name, strlen(if_name));
	
	/* added 09/28/2001 - unset promisc mode */
	if (ioctl (rec, SIOCGIFFLAGS, &if_data) < 0) {
		fprintf (stderr, "couldn`t get flags: %d %s\n", errno, strerror(errno));
		return;
	}
	if (status == SET) if_data.ifr_flags |= IFF_PROMISC;
	if (status == UNSET) if_data.ifr_flags &= ~IFF_PROMISC;
	if (ioctl (rec, SIOCSIFFLAGS, &if_data) < 0) {
		fprintf (stderr, "couldn`t set/unset promisc mode in %d: %d - %s\n", __LINE__, errno, strerror(errno));
	}

}
/* Lazy Ass inc, sorry about the delay :D */


void recvsignal (int sig) {
	logme("received SIGINT, giving up\n");
	promisc_mode(rec, UNSET, if_name);
	if (!outputlog) closelog();
	close (rec);
	exit(0);
	}


void usage (void) {

	fprintf (stderr,"Usage mode:\n");

	fprintf (stderr,"./arpmonitor -i <interface> -s|e -p\n");

	fprintf (stderr,"-i <interface> , interface to check\n");
	fprintf (stderr,"-s or -e , output: <s>yslog or std<e>rr - default: syslog\n");
	fprintf (stderr,"-p setup promisc mode (auto clean on ctrl-C) - default: no promisc mode\n");
	
	exit(0);
}



main.c
/* 
 * arpmonitor	-	log any arp [requests|reply] that comes to the 
 *			local ethernet 
 * 			primary use: debugging.
 * gleicon - 98
 *
 */

#include "main.h"

int rec; /* global fd */
char if_name[256]; /* global iface name, 256 should be enough */
int outputlog=0;

int main (int argc, char **argv) {
	
	int 	len, rsflags, c, promisc_flag=0;
	socklen_t from_len;
	struct 	ifreq		if_data;
	struct 	sockaddr	from;
	char temp_buf[1024];
	char temp_buf2[1024];
	char temp_buf3[1024];

	
	u_int8_t		myMAC[6];
	u_int32_t		ip, sip;



	signal (SIGINT, recvsignal);


	memset (&if_name, 0 , sizeof(if_name));
	memset (&if_data, 0, sizeof (struct ifreq));
	
	while ((c = getopt (argc, argv, OPTSTR)) != EOF) {
	 	switch (c) {
	        	case 'i':
			if (strlen(optarg) < 1) usage();
			strncpy(if_name, optarg, strlen(optarg));
			break;
	        	
			case 's':
			outputlog=0;
			break;
	        	
			case 'e':
			outputlog=1;
			break;
		
	        	case 'p':
			promisc_flag=1;
			break;
			
			case 'h':
			default:
			fprintf (stderr,"This is ArpMonitor\n");
			usage();
			break;
		}									 
	 }
	
	
	
	
	if (getuid ()) {
		fprintf (stderr,"This program must run as root\n");
		exit (0);
		}
	
	if (*(if_name) == 0) {
		fprintf (stderr," -i ????\n");
		usage();
		}
		
	//if ((rec = socket (PF_INET, SOCK_PACKET, htons (ETH_P_ARP))) < 0) {
	if ((rec = socket (PF_INET, SOCK_RAW, htons (ETH_P_ARP))) < 0) {
		syslog(LOG_INFO,"Socket. Aborting.\n");
	        perror("socket");
	        exit (0);
		}
	
	fprintf(stderr, "Iface: %s fd: %d\n", if_name, rec);
	
	strncpy(if_data.ifr_name , if_name, strlen(if_name));

	if (!outputlog)  openlog ("arpmon",LOG_PID, LOG_AUTHPRIV);
	
	logme("Initializing...");
	        

	/* added 09/28/2001 - set promisc mode */

	if (promisc_flag)  promisc_mode(rec, SET, if_name);	

	/* Lazy Ass inc, sorry about the delay :D */


	if (ioctl (rec, SIOCGIFHWADDR, &if_data) < 0) {
		fprintf (stderr,"can't get hardware address\n");
		exit(1);
		}
 	
 	memcpy (myMAC, if_data.ifr_hwaddr.sa_data, 6);
 	
 	sprintf(temp_buf3, "HW address: %s",  hwaddr (myMAC));
	logme (temp_buf3);

 	
 	if (ioctl (rec, SIOCGIFADDR, &if_data) < 0) {
 		fprintf (stderr,"can't get IP address\n");
 		exit(1);
 		}
 	
	memcpy ((void *) &ip, (void *) &if_data.ifr_addr.sa_data + 2, 4);

 	sprintf(temp_buf3, "IP address: %s",  inetaddr(ip));
	logme (temp_buf3);
	
	
	
	if (ioctl (rec, SIOCGIFNETMASK, &if_data) < 0) {
		fprintf (stderr,"can't get NETMASK\n");
		}
	memcpy ((void *) &ip, (void *) &if_data.ifr_netmask.sa_data + 2, 4);

 	sprintf(temp_buf3, "Netmask : %s",  inetaddr(ip));
	logme (temp_buf3);

	if (ioctl (rec, SIOCGIFBRDADDR, &if_data) < 0) {
		fprintf (stderr,"can't get BROADCAST\n");
		}
	
	memcpy ((void *) &ip, (void *) &if_data.ifr_broadaddr.sa_data + 2, 4);
 	
	sprintf(temp_buf3, "Broadcast : %s",  inetaddr(ip));
	logme (temp_buf3);
        
	
	if ((rsflags = fcntl (rec, F_GETFL)) == -1) {
		logme("Fcntl(F_GETFL) . Aborting.");
	        perror ("fcntl F_GETFL");
	        exit (1);
	        }
	                                                                
	if (fcntl (rec, F_SETFL, rsflags | O_NONBLOCK) == -1) {
		logme("Fcntl(F_SETFL) . Aborting.");
	        perror ("fcntl F_SETFL");
	        exit (1);
         	}
	
	        strncpy(from.sa_data, if_name, strlen(if_name));
	        from.sa_family = 1;
	
	for (;;) {
	
		bzero(full_packet, MAX_PACK_LEN);
		
	        // dat magic timeout
		usleep(50);
	        
	        len = recvfrom (rec, full_packet, MAX_PACK_LEN, 0, &from, &from_len);
	       	
		if (len == 0) {
			// waits a bit longer, nothing came from the network
			usleep(100);
			continue;
		} 
	        
	        if (ntohs (arp_pkt.arp_op) == ARPREPLY) {
	        		
	        	memcpy (&ip, arp_pkt.target_ip, 4);
	        	memcpy (&sip, arp_pkt.sender_ip, 4);
	               	sprintf (temp_buf, "[Reply] From: %s [%s] ",inetaddr (sip),hwaddr (arp_pkt.sender_eth));
	        	sprintf (temp_buf2, "To: %s [%s]\n",inetaddr (ip), hwaddr (arp_pkt.target_eth)); 
	        	
			
			if (outputlog) fprintf (stderr,"%s%s\n",temp_buf, temp_buf2);
	        	else syslog (LOG_INFO,"%s%s", temp_buf, temp_buf2);
	        	
	        	}
	        if (ntohs (arp_pkt.arp_op) == ARPREQUEST) {
	        	
	        	memcpy (&ip, arp_pkt.target_ip, 4);
	        	memcpy (&sip, arp_pkt.sender_ip, 4);
	               	sprintf (temp_buf,"[Request] From: %s [%s] ", inetaddr (sip), hwaddr (arp_pkt.sender_eth));
	        	sprintf (temp_buf2,"To: %s [%s]",  inetaddr (ip),  hwaddr (arp_pkt.target_eth));
	        	
			
			if (outputlog) fprintf (stderr,"%s%s\n",temp_buf, temp_buf2);
	        	else syslog (LOG_INFO, "%s%s",temp_buf, temp_buf2);
	        	}
	                                   
	}

	if (promisc_flag) promisc_mode(rec, UNSET, if_name);	
	
	close (rec); 
    	if (!outputlog) closelog();
}


